template/src/api/apiSaga.ts

```diff
@@ -51,8 +51,11 @@ export function* makeApiCall<P>(
     const json = parsedResponse as object;
     return {json, headers};
   } catch (error) {
-    console.error('[makeApiCall] Error:', error.message);
-    yield put(onError(error.message));
+    if (error instanceof Error) {
```

Comment:
Unfortunately, we can't assume that what's being thrown is an `Error`. Anything can be thrown. For example, try this:

```
const someError = "test" 
try {
  throw someError
} catch (error) {
  console.log('Instance ', error, error instanceof Error)
}
```

We can try to do `error as Error` and then process it assuming that it is, but we shouldn't do a conditional that silences unexpected errors.

Reply:
You right, but in case of type assertion we will just log `undefined`

What do You think about this solution for the problem?

```
   if (error instanceof Error) {
      console.error('[makeApiCall] Error:', error.message);
      yield put(onError(error.message));
      return null;
    }

    if (typeof error === 'string') {
      console.error('[makeApiCall] Error:', error);
      yield put(onError(error));
      return null;
    }

    console.error('[makeApiCall] Error:', error);
    yield put(onError('Unknown error'));
    return null;
```

Conde below is inside of catch statement

Reply:
I'm doing something similar for another project right now. So I can share with you what I'm trying out.

1. make the problem visible with adding `"useUnknownInCatchVariables": true,` in `tsconfig`. So it's up to compiler to spot the problem and not to human eyes
2. Try something along this lines
```
  } catch (error) {
    yield* put(fetchWishlistById.Error(getErrorMessage(error)));
  }
}

type ErrorWithMessage = {
  message: string;
};

const isErrorWithMessage = (error: unknown): error is ErrorWithMessage =>
  typeof error === 'object' &&
  error !== null &&
  'message' in error &&
  typeof (error as Record<string, unknown>).message === 'string';

const toErrorWithMessage = (maybeError: unknown): ErrorWithMessage => {
  if (isErrorWithMessage(maybeError)) return maybeError;

  try {
    return new Error(JSON.stringify(maybeError));
  } catch {
    // fallback in case there's an error stringifying the maybeError
    // like with circular references for example.
    return new Error(String(maybeError));
  }
};

const getErrorMessage = (error: unknown) => toErrorWithMessage(error).message;
```

Reply:
This, of course, has to be adapted to your use case, but the idea here is to have some helper methods to handle the "not-so-good" try-catch-error mechanics in JS

Reply:
Sidenote: looks like a blog post here waiting to be written :)

Reply:
Implemented in proposed by Patryk way